internal/retry/observer.go

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package retry
+
+import "github.com/uber-go/tally"
+
+type observer struct {
+	unretryableErrorCounter tally.Counter
+	middlewareErrorCounter  tally.Counter
+	noTimeErrorCounter      tally.Counter
+	maxAttemptsErrorCounter tally.Counter
+	successCounter          tally.Counter
+	attemptCounter          tally.Counter
+}
+
+func newObserver(scope tally.Scope) *observer {
+	return &observer{
+		unretryableErrorCounter: scope.Counter("retry.error.unretryable"),
+		middlewareErrorCounter:  scope.Counter("retry.error.middleware"),
+		noTimeErrorCounter:      scope.Counter("retry.error.notime"),
+		maxAttemptsErrorCounter: scope.Counter("retry.error.maxattempts"),
+		successCounter:          scope.Counter("retry.successes"),
+		attemptCounter:          scope.Counter("retry.attempts"),
```

Comment:
I like the error types that you've laid out here. However, I think we can make these counters more M3-friendly and easier to migrate to pally later.

First, no periods in metric names! Use underscores, which work with pally (and Prometheus).

Second, I think that some of these path segments are actually separate dimensions. I think this will be easier to query if we have

```
name: retry_calls
name: retry_successes
name: retry_failures error: notime
name: retry_failures error: unretryable
...etc...
```

`calls`, `successes`, and `failures` are the same terms used in YARPC and Muttley.